rust-src/bulletproofs/src/set_membership_proof.rs

```diff
@@ -0,0 +1,960 @@
+
+    // Part 2: Verify consistency of t_0
+    // i.e., check 0 = V^z^2 * g^(delta_yz - t_x) * T_1^x * T_2^x^2 * h^(-tx_tilde)
+    let mut delta_minus_tx = delta_yz;
+    delta_minus_tx.sub_assign(&tx);
+    let mut x2 = x; // x^2
+    x2.mul_assign(&x);
+    let mut minus_tx_tilde = tx_tilde;
+    minus_tx_tilde.negate();
+
+    let t0_check_base_points = vec![V.0, v_keys.g, T_1, T_2, v_keys.h];
+    let t0_check_exponents = vec![z2, delta_minus_tx, x, x2, minus_tx_tilde];
+
+    let rhs = multiexp(&t0_check_base_points, &t0_check_exponents);
+    if !rhs.is_zero_point() {
+        return Err(VerificationError::InconsistentT0);
+    }
+
+    // Part 3: Verify inner product
+    // First compute helper variables g_hat, h_prime, and P_prime
+    let g_hat = v_keys.g.mul_by_scalar(&w);
+
+    let y_inv = match y.inverse() {
+        Some(inv) => inv,
+        None => return Err(VerificationError::DivisionError),
+    };
+    let y_inv_n = z_vec(y_inv, 0, n);
+    let mut minus_e_tilde = e_tilde;
+    minus_e_tilde.negate();
+
+    // get exponent for g, i.e., [-z, -z, ..., -z]
+    let mut minus_z = z;
+    minus_z.negate();
+    let mut minus_z_vec = vec![minus_z; n];
+
+    let mut P_prime_exps = Vec::with_capacity(2 * n + 4);
+    P_prime_exps.append(&mut minus_z_vec);
+
+    // compute exponent for h, i.e., z1 + z^2y^-n * s + z^3y^-n
+    for i in 0..n {
+        let mut hexp = z;
+        let mut z2ynisi = z2;
+        z2ynisi.mul_assign(&y_inv_n[i]);
+        z2ynisi.mul_assign(&the_set_vec[i]);
+        hexp.add_assign(&z2ynisi);
+        let mut z3yni = z3;
+        z3yni.mul_assign(&y_inv_n[i]);
+        hexp.add_assign(&z3yni);
+
+        P_prime_exps.push(hexp);
+    }
+
+    // add remaining exponents
+    P_prime_exps.push(tx);
+    P_prime_exps.push(minus_e_tilde);
+    P_prime_exps.push(C::Scalar::one());
+    P_prime_exps.push(x);
+
+    let P_prime_bases = vec![g_hat, v_keys.h, A, S];
+
+    // Finally verify inner product
+    let ip_verification = verify_inner_product_with_scalars(
+        transcript,
+        gens,
+        &y_inv_n,
+        &P_prime_bases,
+        &P_prime_exps,
+        &g_hat,
+        &proof.ip_proof,
+    );
+
+    if !ip_verification {
+        return Err(VerificationError::IPVerificationError);
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pairing::bls12_381::G1;
+    type SomeCurve = G1;
+
+    /// This function verifies a set membership proof, i.e. a proof of knowledge
+    /// of value v that is in a set S and that is consistent
+    /// with a commitment V to v. The arguments are
+    /// - S - the set as a vector
+    /// - V - commitment to v
+    /// - proof - the set membership proof to verify
+    /// - gens - generators containing vectors G and H both of length at least
+    ///   |S| (bold g,h in bluepaper)
+    /// - v_keys - commitment keys B and B_tilde (g,h in bluepaper)
+    #[allow(non_snake_case)]
+    pub fn verify_naive<C: Curve>(
```

Comment:
Please describe also why this is here. What is "naive" about it.

Reply:
I suggest to remove the function. The reason we had is that it calls the existing inner product verification black box, while the other ones requires the optimized inner product verification with scalars. If you are OK with adding the new inner product verification, there is not reason to have this "naive" function here.